modules/platforms/dotnet/Apache.Ignite.Core/Impl/Cache/CacheImpl.cs

```diff
@@ -817,36 +817,129 @@ public Task RemoveAllAsync()
         /** <inheritDoc /> */
         public int GetLocalSize(params CachePeekMode[] modes)
         {
-            return Size0(true, modes);
+            return Size0<int>(true, null, modes);
         }
 
         /** <inheritDoc /> */
         public int GetSize(params CachePeekMode[] modes)
         {
-            return Size0(false, modes);
+            return Size0<int>(false, null, modes);
         }
 
         /** <inheritDoc /> */
         public Task<int> GetSizeAsync(params CachePeekMode[] modes)
         {
-            var modes0 = IgniteUtils.EncodePeekModes(modes);
+            return SizeAsync0<int>(null, modes);
+        }
+
+        /** <inheritDoc /> */
+        public long GetSizeLong(params CachePeekMode[] modes)
+        {   
+            return Size0<long>(false, null, modes);
+        }
+
+        /** <inheritDoc /> */
+        public long GetSizeLong(int partition, params CachePeekMode[] modes)
+        {
+            return Size0<long>(false, partition, modes);
+        }
 
-            return DoOutOpAsync<int>(CacheOp.SizeAsync, w => w.WriteInt(modes0));
+        /** <inheritDoc /> */
+        public Task<long> GetSizeLongAsync(params CachePeekMode[] modes)
+        {
+            return SizeAsync0<long>(null, modes);
+        }
+
+        /** <inheritDoc /> */
+        public Task<long> GetSizeLongAsync(int partition, params CachePeekMode[] modes)
+        {
+            return SizeAsync0<long>(partition, modes);
+        }
+
+        /** <inheritDoc /> */
+        public long GetLocalSizeLong(params CachePeekMode[] modes)
+        {
+            return Size0<long>(true, null, modes);
+        }
+
+        /** <inheritDoc /> */
+        public long GetLocalSizeLong(int partition, params CachePeekMode[] modes)
+        {
+            return Size0<long>(true, partition, modes);
         }
 
         /// <summary>
         /// Internal size routine.
         /// </summary>
         /// <param name="loc">Local flag.</param>
+        /// <param name="part">Partition number</param>
         /// <param name="modes">peek modes</param>
         /// <returns>Size.</returns>
-        private int Size0(bool loc, params CachePeekMode[] modes)
+        private T Size0<T>(bool loc, int? part, params CachePeekMode[] modes)
         {
             var modes0 = IgniteUtils.EncodePeekModes(modes);
 
-            var op = loc ? CacheOp.SizeLoc : CacheOp.Size;
+            var op = loc ? 
+                (typeof(T) == typeof(long) ? CacheOp.SizeLongLoc : CacheOp.SizeLoc) : 
+                (typeof(T) == typeof(long) ? CacheOp.SizeLong : CacheOp.Size);
+
+            long ret;
+            
+            if (typeof(T) == typeof(long))
```

Comment:
Too complicated - we have to check for types, then convert the type.
Let's have two separate methods for `long` and `int`, they have very little in common.

Reply:
Yes, it's seems to be a little bit overengineered.